src/metrics/handle.rs

```diff
@@ -0,0 +1,257 @@
+            (4.0, AtomicU32::new(0)),
+            (8.0, AtomicU32::new(0)),
+            (16.0, AtomicU32::new(0)),
+            (32.0, AtomicU32::new(0)),
+            (32.0, AtomicU32::new(0)),
+            (128.0, AtomicU32::new(0)),
+            (256.0, AtomicU32::new(0)),
+            (512.0, AtomicU32::new(0)),
+            (1024.0, AtomicU32::new(0)),
+            (2048.0, AtomicU32::new(0)),
+            (4096.0, AtomicU32::new(0)),
+            (f64::INFINITY, AtomicU32::new(0)),
+        ]);
+        Self {
+            buckets,
+            count: AtomicU32::new(0),
+            sum: AtomicU64::new(0),
+        }
+    }
+
+    pub(crate) fn record(&mut self, value: f64) {
+        let mut prev_bound = f64::NEG_INFINITY;
+        assert!(self.buckets.len() == 22);
+        for (bound, bucket) in self.buckets.iter_mut() {
+            if value > prev_bound && value <= *bound {
+                bucket.fetch_add(1, Ordering::Relaxed);
+                break;
+            } else {
+                prev_bound = *bound;
+            }
+        }
+
+        self.count.fetch_add(1, Ordering::Relaxed);
+        let _ = self
+            .sum
+            .fetch_update(Ordering::AcqRel, Ordering::Relaxed, |cur| {
+                let next_sum = f64::from_bits(cur) + value;
+                Some(next_sum.to_bits())
+            });
+    }
+
+    pub fn count(&self) -> u32 {
+        self.count.load(Ordering::Relaxed)
+    }
+
+    pub fn sum(&self) -> f64 {
+        f64::from_bits(self.sum.load(Ordering::Relaxed))
+    }
+
+    pub fn buckets(&self, buf: &mut Vec<(f64, u32)>) {
+        for (k, v) in self.buckets.iter() {
+            buf.push((*k, v.load(Ordering::Relaxed)));
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Counter {
+    inner: AtomicU64,
+}
+
+impl Counter {
+    pub(crate) fn with_count(count: u64) -> Self {
+        Self {
+            inner: AtomicU64::new(count),
+        }
+    }
+
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: AtomicU64::new(0),
+        }
+    }
+
+    pub(crate) fn record(&mut self, value: u64) {
+        self.inner.fetch_add(value, Ordering::Relaxed);
+    }
+
+    pub fn count(&self) -> u64 {
+        self.inner.load(Ordering::Relaxed)
+    }
+}
+
+#[derive(Debug)]
+pub struct Gauge {
+    inner: AtomicU64,
+}
+
+impl Gauge {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: AtomicU64::new(0),
+        }
+    }
+
+    pub(crate) fn record(&mut self, value: GaugeValue) {
+        // Because Rust lacks an atomic f64 we store gauges as AtomicU64
+        // and transmute back and forth to an f64 here. They have the
+        // same size so this operation is safe, just don't read the
+        // AtomicU64 directly.
+        let _ = self
+            .inner
+            .fetch_update(Ordering::AcqRel, Ordering::Relaxed, |cur| {
+                let val = value.update_value(f64::from_bits(cur));
+                Some(val.to_bits())
+            });
```

Comment:
Since we are using this "atomic-f64-in-u64" pattern in two places, I would suggest making an `AtomicF64` wrapper to unify and make the pattern explicit.

Reply:
Agreed.